Add WRot multiplication support.

pchote opened the pull request:
This PR adds the plumbing to convert from a quaternion back to a `WRot`, and uses the quaternions to implement multiplication over rotations. This plumbing will be used by my turret facing rework to finally fix turret/barrel/muzzle positioning and orientation on slopes.

The first commit implements the low level maths - see https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles for more details. #17796 also had an implementation for this, but it relied on floating point calculations that aren't suitable for simulation code.

The second commit changes the voxel rendering to take a single resolved rotation instead of an enumerable that it applies in order. This is technically a step backwards because converting back to a single `WRot` is slower and less precise, but we win significantly in reduced complexity and consistency by being able to use the same implementation for simulation and rendering. We may regret this, but I think its the right choice at least in the short term.

The key tests here are to make sure that the PR maintains the same behaviour as current bleed for turret and muzzle positions on voxel units. You can add `orientation = new WRot(WAngle.Zero, new WAngle(128), WAngle.Zero);` before the `Facing = oldFacing = ...` line in the `Mobile` ctor to check how (broken) it looks for inclined units.

You can do something like
```csharp
for (var i = 0; i <= 1024; i++)
	System.Console.WriteLine("ArcSin(Sin({0})) = {1}", i, WAngle.ArcSin(new WAngle(i).Sin()));
```


to verify that the new `WAngle.ArcSin` and `WAngle.ArcCos` behave as expected - but note that we do expect some disagreement due to (a) the limited integer precision, and (b) the periodic domain of trig functions causing the round-tripped value to be reflected/offset. This precision limitation makes it difficult to write any proper unit tests for this.

While this isn't strictly needed to fix the facing &rarr; WAngle conversion and world &rarr; body origin in `Turreted`, this allows us to do it properly instead of adjusting incorrect code in the short term only to replace it later.

pchote committed:
Add WRot.Rotate to allow rotations to be combined.

tovl commented on OpenRA.Game/WRot.cs:
```diff
@@ -29,6 +29,27 @@ public struct WRot : IEquatable<WRot>
 		public static WRot operator -(WRot a, WRot b) { return new WRot(a.Roll - b.Roll, a.Pitch - b.Pitch, a.Yaw - b.Yaw); }
```
The + and - operators should really be removed, since they are not valid operators for 3d-rotations. They only accidentally give an answer that seems to make sense when pitch and roll are zero.

pchote replied:
Good idea, and their use is exactly why things break so badly when introducing slopes.

I'll add this to my list of things to do in the next PR, which removes their from turrets etc.

tovl commented on OpenRA.Game/WRot.cs:
```diff
@@ -29,6 +29,27 @@ public struct WRot : IEquatable<WRot>
 		public static WRot operator -(WRot a, WRot b) { return new WRot(a.Roll - b.Roll, a.Pitch - b.Pitch, a.Yaw - b.Yaw); }
 		public static WRot operator -(WRot a) { return new WRot(-a.Roll, -a.Pitch, -a.Yaw); }
 
+		public static WRot operator *(WRot a, WRot b)
```
We should be really careful when applying this operator, since multiplying rotations is non-commutative! I think at least a comment to that effect here would be in order because I can imagine a lot of potential frustration when people aren't aware of that when trying to use it. We should also double check the applications in the second commit.

pchote replied:
The second commit accounts for the non-commutativity, but good idea to add a comment here.

tovl commented on OpenRA.Game/WRot.cs:
```diff
@@ -29,6 +29,27 @@ public struct WRot : IEquatable<WRot>
 
+		public static WRot operator *(WRot a, WRot b)
+		{
+			if (a == Zero)
+				return b;
```
This sounds really odd when you think about it, which makes me think that calling this `Zero` was a mistake. But not sure what to do about it now.

pchote replied:
Good point. Any objection to `WRot.None`? `WRot.Identity` is a bit less obvious.

There are hardly any uses, so I can scope creep this PR to include a rename.

abcdefg30 replied:
`WRot.None` sounds good to me. :+1:

tovl commented:
We convert the Euler angles to Quaternions and from quaternions to a rotation matrix every time we use the rotation for rendering, every time we apply the rotation to a vector and now also convert to quaternions and back every time we want to combine two rotations. Basically, we do these conversions every time we actually want to do anything with them. I'm starting to wonder if there would be any advantage in storing the quarternions internally so we only do the conversion once upon creation?

pchote commented:
That thought occurred to me too, but I dismissed it because I figured it would conflict with the places where we access or set roll/pitch/yaw directly. But, as you point out, most of our actual uses of `WRot` are from matrix/quaternion form so I think you may be right - we can always calculate the roll/pitch/yaw when needed.

This will also avoid some of the precision issues I pointed out in the PR description.

pchote commented:
~However... this conversion depends on us first removing the `operator +` and `operator -` implementations, so I would prefer to leave this to a third PR after the second one sorts out turrets etc.~

pchote committed:
Rename WRot.Zero to WRot.None.

pchote committed:
Pass pre-combined rotations into the rendering code.

pchote commented:
Updated:
* Renamed `WRot.Zero` to `WRot.None`
* Replaced the multiplication operator with an explicit `.Rotate` method to make it clear which is operating on the other and to match `WVec`.

pchote commented:
Added another commit to store both Euler angles and quaternion components in WRot.

This should give us the best of both approaches, for only a small memory tradeoff.

pchote committed:
Store quaternion components directly.

This avoids precision loss when combining rotations.
The equivalent Euler angles are calculated for external use
but the quaternion components are preferred for any further
internal calculations.

teinarss reviewed (approved):
Code looks fine and couldn't find any regression ingame.

tovl commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d8b1fa222e34b9ee5150f8f2f8c9f1a8924063dd...9e7e549295f834b04bff24779e6150cca27de9be
